### PR TITLE
Optical flow fusion control improvements

### PR DESCRIPTION
**Pull request by mhkabir**
Brings in some fixes and improvements in optical flow fusion control. Includes https://github.com/PX4/ecl/pull/502 as it was needed for flight testing. Most of these corner cases are more visible on platforms running optical flow algorithms offboard. (e.g Aero,  Snapdragon Flight)

Functional changes : 
1. Add protection against large dts, which can put the fusion control into a bad state (if fusion stops, it never starts again) due to float overflows and unsigned integer checks.
 (https://github.com/PX4/ecl/blob/master/EKF/control.cpp#L496)
2. Remove unnecessary check for tilt angle when popping new data from flow buffer. There is a similar check downstream, and this one was interfering with correct fusion inhibition detection.
3. Handle flow data timeouts correctly (fusion timeouts != data stopped)
4. Remove state predictor alignment which doesn't seem to be used for any of the other aiding sources.  
5. Add 2 minor `ECL_INFO` statements which go a long way in figuring out what went wrong in logs.

Log :
https://logs.px4.io/plot_app?log=a13404d3-078e-4f04-94e0-19a1f287bc72

### CODE REVIEW

**Comment by mhkabir** [comment]
More testing revealed an additional corner case which can cause divergence. Let me investigate further.

**Comment by xdwgood** [comment]
screenshot from 2018-09-06 10-54-14

**Comment by mhkabir** [comment]
@xdwgood Additional context?

**Comment by xdwgood** [comment]
@mhkabir  I use the log you provided in this PR.

**Comment by mhkabir** [comment]
@priseborough Can you please have a look here? The fixes here are crucial and necessary, and some help figuring out the last 5% would be appreciated, thanks!

**Comment by priseborough** [comment]
OK will do

**Comment by mhkabir** [comment]
@priseborough Any luck?

**Comment by mhkabir** [comment]
@priseborough Ping.

**Comment by priseborough** [comment]
@mhkabir I assume the last 5% you are referring to is the position divergence at 375 seconds in the log https://review.px4.io/plot_app?log=488ad808-7886-4372-a0da-7f51a04be2fe .

Unfortunately there is nothing that stands out in the data itself to explain what happened so I tried running a replay using upstream master but was unable to due to the large number of data dropouts and or the mismatched topics. Where did you fork form upstream for PX4/Firmware?

Resolving this is going to require a data log suitable for replay. If replacing the SD card with a better quality one or reformatting the one you doesn't work, you may have to set SDLOG_PROFILE = 2

**Comment by priseborough** [comment]
@mhkabir Are you able to provide a log suitable for replay? Without that, diagnosing the divergence seen in your testing is impractical.

**Comment by mhkabir** [comment]
@RomanBapst @DanielePettenuzzo Could you guys give this a test with your setup after rebasing on master?

**Comment by priseborough** [comment]
@mhkabir The reported position divergence that may have been fixed by the covariance reset fixes that have gone in recently. It would be worth rebasing and retesting.

If the divergence issue is still present, then a replay log is required to perform the required diagnostics.

**Comment by mhkabir** [comment]
I'll rebase.

**Comment by RomanBapst** [comment]
@mhkabir Can you please rebase this on master? I have a vehicle here ready for optical flow and I can provide a replay log so that we can continue with this.

**Comment by mhkabir** [comment]
@RomanBapst @DanielePettenuzzo Rebased, please test!

**Comment by priseborough** [comment]
@mhkabir Have you done your own testing on this PR ?

File: EKF/estimator_interface.cpp [hunk 1]
```diff
@@ -366,31 +366,28 @@ void EstimatorInterface::setOpticalFlowData(uint64_t time_usec, flow_message *fl
 		// of min arrival interval because too much data is being lost
 		float delta_time = 1e-6f * (float)flow->dt;
 		float delta_time_min = 5e-7f * (float)_min_obs_interval_us;
-		bool delta_time_good = delta_time >= delta_time_min;
+		float delta_time_max = 0.1f;
```
Comment (priseborough) [inline EKF/estimator_interface.cpp]: Is this value based on testing or set above what current supported sensors output. It would prevent use of a 10Hz sensor.

**Comment by priseborough** [comment]
The change to reset logic in https://github.com/PX4/ecl/pull/505/commits/377829b434a7c8b03f0f5d43ac1996f115dcd405 requires careful testing.

File: EKF/estimator_interface.cpp (hunk 1 above)
Comment (dagar) [reply EKF/estimator_interface.cpp]: > Is this value based on testing or set above what current supported sensors output. It would prevent use of a 10Hz sensor.

@mhkabir ?

Comment (mhkabir) [reply EKF/estimator_interface.cpp]: I set 10Hz fairly arbitrarily. I can increase to 5Hz minimum. Will this be fine? 

The PMW3901 runs at 40Hz. I'm not sure about PX4Flow.

**Comment by mhkabir** [comment]
@jkflying @RomanBapst Would you be able to do some more structured testing to the reset logic. These should be carried out with logging from boot, and optical flow as the only enabled sensor in `EKF2_AID_MASK`: 

1. Shake the vehicle fast, which should trigger the bad data quality checks (https://github.com/PX4/ecl/pull/505/files#diff-4cac31c333867233bb8a08de1e812a9aR424). Fusion should recover once the shaking is stopped.

2. Disconnect the sensor physically or stop the driver from console while fusion is active. This should trigger the data loss checks (https://github.com/PX4/ecl/pull/505/files#diff-4cac31c333867233bb8a08de1e812a9aR490). 

In both of these tests, local position should go invalid with appropriate messages in the console, and then recover once initial conditions are restored.

**Comment by priseborough** [comment]
Any update on completing testing? I do not support merging this until the testing is complete.

**Comment by mhkabir** [comment]
@jkflying @RomanBapst

File: EKF/control.cpp [hunk 2]
```diff
@@ -419,45 +418,34 @@ void Ekf::controlOpticalFlowFusion()
...
-			       _time_last_of_fuse = 0;
-
-			} else if ((_time_last_imu - _time_last_of_fuse) > (uint64_t)_params.reset_timeout_max) {
-				_control_status.flags.opt_flow = false;
-
-			}
+		// Handle cases where we are using optical flow but are no longer able to because
+		// its use has been inhibited.
+		if (_control_status.flags.opt_flow && _inhibit_flow_use) {
+			ECL_WARN("Optical flow data quality poor - stopping use");
```
Comment (priseborough) [inline EKF/control.cpp]: _inhibit_flow_use can toggle on and off as the vehicle goes between hover and moving if it is also using GPS. Because of this this message may flood the interface.

Comment (priseborough) [reply EKF/control.cpp]: Also this ability to switch flow use on and off depending on vehicle movement (when it is also using GPS) is a feature and does not necessarily indicate a fault or bad flow data, just an operating condition where it is better to not use flow.

File: EKF/control.cpp [hunk 3]
```diff
@@ -119,8 +119,7 @@ void Ekf::controlFusionModes()
 	// We don't fuse flow data immediately because we have to wait for the mid integration point to fall behind the fusion time horizon.
 	// This means we stop looking for new data until the old data has been fused.
 	if (!_flow_data_ready) {
-		_flow_data_ready = _flow_buffer.pop_first_older_than(_imu_sample_delayed.time_us, &_flow_sample_delayed)
-				   && (_R_to_earth(2, 2) > _params.range_cos_max_tilt);
+		_flow_data_ready = _flow_buffer.pop_first_older_than(_imu_sample_delayed.time_us, &_flow_sample_delayed);
```
Comment (priseborough) [inline EKF/control.cpp]: I don't see downstream where the use of flow data when in air is stopped if the vehicle is tilted excessively other than indirectly through loss of range fusion which them after several seconds causes terrain to go invalid.

I think this change will allow use of flow data in flight when excessively tilted.

**Comment by priseborough** [comment]
@RomanBapst and I walked through the changes and found  two issues that need to be fixed - see comments.

**Comment by priseborough** [comment]
@mhkabir Any progress?

**Comment by mhkabir** [comment]
Working on it. I have a flow setup going again with this new sensor: https://github.com/PX4/Firmware/pull/12148/. Will keep you posted.

File: EKF/control.cpp (hunk 3 above)
Comment (mhkabir) [reply EKF/control.cpp]: This is here : https://github.com/PX4/ecl/pull/505/files#diff-4cac31c333867233bb8a08de1e812a9aR370
